File: frontend/src/types/offer.ts
// Offer Tracker — mirrors backend app/schemas/offer.py

export type OfferStatus =
  | 'draft'
  | 'submitted'
  | 'countered'
  | 'accepted'
  | 'rejected'
  | 'withdrawn'
  | 'expired'

export interface PropertyOffer {
  id: string
  saved_property_id: string
  /** Serialized Decimal — parse with parseFloat for display. */
  amount: string | number
  counter_amount: string | number | null
  status: OfferStatus
  offer_date: string
  expires_at: string | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface PropertyOfferCreate {
  amount: number
  status?: OfferStatus
  counter_amount?: number | null
  offer_date?: string | null
  expires_at?: string | null
  notes?: string | null
}

export interface PropertyOfferUpdate {
  amount?: number
  status?: OfferStatus
  counter_amount?: number | null
  offer_date?: string | null
  expires_at?: string | null
  notes?: string | null
}

File: backend/app/schemas/offer.py
"""Schemas for PropertyOffer CRUD."""

from datetime import datetime
from decimal import Decimal
from typing import Annotated

from pydantic import BaseModel, BeforeValidator, Field

from app.models.offer import OfferStatus


def _accept_bare_date(v: object) -> object:
    """The frontend's <input type=date> submits ``YYYY-MM-DD``; widen it to a
    midnight datetime so Pydantic's datetime parser accepts it."""
    if isinstance(v, str) and len(v) == 10 and v.count("-") == 2:
        return f"{v}T00:00:00"
    return v


FlexibleDatetime = Annotated[datetime, BeforeValidator(_accept_bare_date)]


class OfferCreate(BaseModel):
    amount: Decimal = Field(..., gt=0, le=Decimal("9999999999.99"))
    status: OfferStatus = OfferStatus.SUBMITTED
    counter_amount: Decimal | None = Field(None, gt=0, le=Decimal("9999999999.99"))
    offer_date: FlexibleDatetime | None = None
    expires_at: FlexibleDatetime | None = None
    notes: str | None = Field(None, max_length=4000)


class OfferUpdate(BaseModel):
    """All fields optional — partial updates only."""

    amount: Decimal | None = Field(None, gt=0, le=Decimal("9999999999.99"))
    status: OfferStatus | None = None
    counter_amount: Decimal | None = Field(None, gt=0, le=Decimal("9999999999.99"))
    offer_date: FlexibleDatetime | None = None
    expires_at: FlexibleDatetime | None = None
    notes: str | None = Field(None, max_length=4000)


class OfferOut(BaseModel):
    id: str
    saved_property_id: str
    amount: Decimal
    counter_amount: Decimal | None
    status: OfferStatus
    offer_date: datetime
    expires_at: datetime | None
    notes: str | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
